Deprecate did-get-response-details and did-get-redirect-request

nitsakh opened the pull request:
Resolves #12597.
The two events mentioned are being deprecated and will completely go away in chromium 66 upgrade.

nitsakh commented on lib/common/api/deprecate.js:
```diff
@@ -46,6 +46,26 @@ deprecate.log = (message) => {
+    warned = true
+    // Event name starts with a -, which means it's an internal event.
+    // For deprecation, that means the event cannot be used anymore.
+    if (!newName.startsWith('-')) { deprecate.warn(`'${oldName}' event`, `'${newName}' event`) } else { deprecate.log(`'${oldName}' event has been deprecated.`) }
```
:-( Linter fix made this line long.

alexeykuzmin replied:
You mean you cannot use line breaks here?

nitsakh replied:
I tried having if and else on separate lines. But the linter auto fixed to this. Maybe I can trick it somehow, but it's probably fine.

alexeykuzmin replied:
I thought we have only one linter for all js files.
Let's ignore the linter if it suggests this nonsense ), and write it like that:

```js
if () {
  ...
} else {
  ...
}
```

codebytere commented on atom/browser/api/atom_api_web_contents.cc:
```diff
@@ -860,7 +860,7 @@ void WebContents::DidGetResourceResponseStart(
        details.resource_type == content::RESOURCE_TYPE_SUB_FRAME))
     return;
-  Emit("did-get-response-details",
+  Emit("-did-get-response-details",
```
this seems to follow the comments for event deprecation in the method, but instinctively i feel like deprecations should maintain backwards compatibility and this change will break existing listeners 🤔

ckerr replied:
I agree, and that kind of goes against the purpose of the deprecation guard.

Maybe we should emit _both_ events during the interim deprecated phase, then remove the nonprefixed one when we remove the public facing one.

poiru replied:
The `deprecate.event(...)` additions to `api-web-contents.js` should cause the old event to still fire.

codebytere replied:
oh yep, i see that now! in that case, this looks good to go

nitsakh commented:
@codebytere I updated the guest view files. For the tests, I updated the events, instead of removing them completely, just so that we are sure they still work.

alexeykuzmin commented on lib/common/api/deprecate.js:
```diff
@@ -46,6 +46,26 @@ deprecate.log = (message) => {
+    if (fn != null) {
+      fn.apply(this, arguments)
+    } else {
+      this.emit.apply(this, [oldName].concat(args))
```
`this.emit.call(this, oldName, ...args)`

nitsakh replied:
Why is call better than apply?

alexeykuzmin replied:
@nitsakh 
You probably somehow saw the very first version of my suggestion )
`this.emit(oldName, ...args)` is much better alternative here.

----

_But if you're curious_
> Why is `call` better than `apply`?

In general there's no difference.
But `this.emit.call(this, oldName, ...args)` is better than `this.emit.apply(this, [oldName].concat(args))` because the second option makes you _think_ more and that's not what you should be doing while reading code =)

alexeykuzmin commented on lib/common/api/deprecate.js:
```diff
@@ -46,6 +46,26 @@ deprecate.log = (message) => {
 }
 
+// Deprecate an event.
+deprecate.event = (emitter, oldName, newName, fn) => {
```
You never call `deprecate.event` with `fn` argument. Why is it needed?

nitsakh replied:
I don't! But we can give `deprecate.event` a function to run. So, this just makes it general.
We've done this in the past, for example:
```
deprecate.event(app, 'activate-with-no-open-windows', 'activate', function (event, hasVisibleWindows) {
  if (!hasVisibleWindows) {
    return this.emit('activate-with-no-open-windows', event)
  }
})
```

codebytere replied:
@alexeykuzmin that was how it was originally written for extensibility by kevin :)

alexeykuzmin replied:
@codebytere Yeah, I saw it in the removed code. But I see no reasons to keep it.
We can always add it later if needed.

nitsakh replied:
Done. Removed.

alexeykuzmin commented on lib/common/api/deprecate.js:
```diff
@@ -46,6 +46,26 @@ deprecate.log = (message) => {
+    // There are no listeners for this event
+    if (this.listenerCount(oldName) === 0) { return }
+    // noDeprecation set or if user has already been warned
+    if (warned || process.noDeprecation) { return }
```
Can `process.noDeprecation` be changed in runtime?
If no, then maybe `deprecate.event()` shouldn't do anything if the flag is set?

nitsakh replied:
As far as I know, setting `process.noDeprecation` at runtime should work.

alexeykuzmin commented on lib/common/api/deprecate.js:
```diff
@@ -46,6 +46,26 @@ deprecate.log = (message) => {
+    warned = true
+    // Event name starts with a -, which means it's an internal event.
+    // For deprecation, that means the event cannot be used anymore.
+    if (!newName.startsWith('-')) {
```
The comment says "Event name starts with a -", but the code says that it doesn't: `if (!newName.startsWith('-'))`.
Who's lying? )

nitsakh replied:
No one or both. 😄
I will update.

alexeykuzmin commented on lib/common/api/deprecate.js:
```diff
@@ -46,6 +46,26 @@ deprecate.log = (message) => {
+    // noDeprecation set or if user has already been warned
+    if (warned || process.noDeprecation) { return }
+    warned = true
+    // Event name does not start with a -, so it's not an internal event.
```
I didn't mean to criticize this piece too hard, but now this line just sounds too complicated.
I _think_ that it can become clearer if it is rewritten as something like this:

```js
const isInternalEvent = newName.startsWith('-')
if (isInternalEvent) {
  // It kinda ok I guess, so only a log message here.
  deprecate.log(...)
} else {
  // The event cannot be used anymore.
  deprecate.warn(...)
}
```

alexeykuzmin commented on lib/common/api/deprecate.js:
```diff
@@ -55,12 +55,13 @@ deprecate.event = (emitter, oldName, newName) => {
-    } else {
+    const isInternalEvent = newName.startsWith('-')
+    if (isInternalEvent) {
+      // The event cannot be use anymore. Log that.
```
"used" )

alexeykuzmin commented:
@nitsakh Thank you! =)

deepak1556 commented on docs/api/webview-tag.md:
```diff
@@ -679,6 +679,7 @@ Returns:
 `status` indicates socket connection to download the resource.
 
+**Deprecated**: This event has been deprecated. Use the [`webRequest`](web-request.md) module which provides similar navigation details on a subscription basis.
 ### Event: 'did-get-redirect-request'
```
Should also mark this event as `Deprecated`

nitsakh commented:
Missed updating the docs to indicate deprecation for other event. Updated that.
